### DISCUSSION TITLE
The value I save to storage with persist is delayed when the page refreshes.

### DISCUSSION THREAD

**Discussion by aydinvivik** [pmndrs/zustand#2450] [Bug report]
## Summary
When I refresh the page with the value I saved to localstorage with persist, it takes the old value for 1 second and then shows the new value. Is there any way to avoid this 1 second delay?

My example code is below;

use-store.js
```
'use client';

import { useState, useEffect } from 'react';

const useStore = (store, callback) => {
  const result = store(callback);
  const [data, setData] = useState();

  useEffect(() => {
    setData(result);
  }, [result]);

  return data;
};

export default useStore;
```

use-template.js
```
import { create } from 'zustand';
import { persist } from 'zustand/middleware';

const initialStates = {
  templateStatus: false,
};

const useTemplate = create(
  persist(
    (set) => ({
      ...initialStates,
      setTemplateStatus: () => {
        set((state) => ({ templateStatus: !state.templateStatus }));
      },
    }),
    {
      name: 'template-status',
    }
  )
);

export default useTemplate;
```

component.js
```
'use client';

import useStore from './store/use-store';
import useTemplate from './store/use-template';

export default function Home() {
  const templateStatus = useStore(
    useTemplate,
    (state: any) => state.templateStatus
  );
  const setTemplateStatus = useTemplate((state) => state.setTemplateStatus);
  return (
    <main className="flex min-h-screen flex-col items-center justify-between p-24">
      <div className="mb-32 grid text-center lg:max-w-5xl lg:w-full lg:mb-0 lg:grid-cols-4 lg:text-left">
        <p>template status: {templateStatus ? 'true' : 'false'}</p>
        <button onClick={setTemplateStatus}>change template status</button>
      </div>
    </main>
  );
}
```

## Link to reproduction
https://stackblitz.com/edit/stackblitz-starters-jjsutv?file=app%2Fpage.tsx

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [ ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [ ] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
Related with #2446?

**Reply by aydinvivik** [reply]
Yes, I thought I wrote that in the wrong place, I can delete the old one.

**Comment by aydinvivik** [comment]
I don't want to go back to redux but I don't understand how to solve such a problem.

**Reply by dbritto-dev** [reply]
@aydinvivik have you ever read our nextjs guide?

**Reply by aydinvivik** [reply]
yes i added the provider but again the result is the same. i think i am missing something somewhere?

**Reply by dbritto-dev** [reply]
@aydinvivik yeah, you need to skip hydration and hydrate using useEffect

**Reply by dbritto-dev** [reply]
@aydinvivik here you go https://docs.pmnd.rs/zustand/integrations/persisting-store-data#skiphydration

**Reply by aydinvivik** [reply]
I tried but the result is the same.

https://stackblitz.com/edit/stackblitz-starters-jjsutv?file=app%2Fpage.tsx

**Reply by aydinvivik** [reply]
I have the same problem when I do the example in the documentation?

https://stackblitz.com/edit/stackblitz-starters-1mbm3v?file=app%2Fstores%2Fcounter-store.ts

**Reply by dbritto-dev** [reply]
@aydinvivik I'll create an example for you

**Reply by aydinvivik** [reply]
I would be very grateful if you could help me solve this problem.

**Reply by aydinvivik** [reply]
@dbritto-dev Hello, I'm taking your time but I wanted to know if there is any progress for example?

**Reply by dbritto-dev** [reply]
@aydinvivik here you go -> https://stackblitz.com/edit/stackblitz-starters-hwoh91

**Reply by aydinvivik** [reply]
@dbritto-dev Thank you, but I think loading is hiding the problem here, if there is no loading I have the same situation again.

**Reply by dbritto-dev** [reply]
@aydinvivik you can't avoid that, reading from local storage is an async operation and you need to make sure the data is available before showing something

**Reply by dbritto-dev** [reply]
btw, you shouldn't hydrate/rehydrate or try to read something on server client when you are working on next or any ssr. we mentioned that on our nextjs guide.

**Reply by aydinvivik** [reply]
I understand, thank you for your feedback.

**Reply by overvi** [reply]
```javascript
import ThemeProvider from "@/providers/ThemeProvider";
import type { Metadata } from "next";
import { Inter } from "next/font/google";
import "./globals.css";

const InterF = Inter({
  variable: "--font-inter",
  subsets: ["cyrillic"],
});

export const metadata: Metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en" suppressHydrationWarning>
      <head>
        <script
          dangerouslySetInnerHTML={{
            __html: `
                (function() {
              document.documentElement.style.setProperty('--base-font' , JSON.parse(localStorage.getItem("font-store")).state.fontSize)
                })();
              `,
          }}
        ></script>
      </head>

      <body
        className={`${InterF.variable} font-inter antialiased bg-main min-h-screen`}
      >
        <ThemeProvider>{children}</ThemeProvider>
      </body>
    </html>
  );
}
```

You can get the value directly in the head tag so it will load before the html

**Reply by aydinvivik** [reply]
Thank you @overvi I will try.
